### ISSUE TITLE
Overfitting sanity check slow

### ISSUE THREAD

**Issue by westand** [deeplearning4j/deeplearning4j#2116]
Per [http://cs231n.github.io/neural-networks-3/#sanitycheck] I am trying to sanity check my setup and DataIterator. I should be able to get my loss to 0 by overfitting very few training examples. I am piping in 16 training examples to the below network. It is taking nearly 1000 epochs to get the loss down to 0.001. 

This is absurdly long, right? Where might this indicate a problem lies? For context, I am classifying short words into 2 classes using sequences of one-hot character vectors. I've tried to turn off all regularization and optimization here for the sake of the sanity-check. 

<B>The network itself:</B>
<pre>int inputCols = PeriodIterator.CHARSET.size();
int lstmLayerSize = 256;
int nOut = 2;
int batchSize = 64; 
int nEpochs = 5000; 
int charLen = 5;
          
 MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
		.seed(123)
		.learningRate(0.1)
		.regularization(false)
		.list()
		.layer(0, new GravesBidirectionalLSTM.Builder()
				.nIn(inputCols)
				.nOut(lstmLayerSize)
				.build())
		.layer(1, new RnnOutputLayer.Builder()
				.nIn(lstmLayerSize)
				.nOut(nOut)
				.build())
		.backprop(true)
		.pretrain(false)
		.build();
		
DataSetIterator train = new PeriodIterator(batchSize, charLen, true);
DataSetIterator test = new PeriodIterator(batchSize, charLen, false);</pre>

<B>The meat of the iterator:</B>
<pre>INDArray features = Nd4j.create(new int[]{words.size(), CHARSET.size(), truncLen}, 'f');
INDArray labels = Nd4j.create(new int[]{words.size(), 2, truncLen}, 'f');
INDArray featuresMask = Nd4j.zeros(new int[]{words.size(), truncLen}, 'f');
INDArray labelsMask = Nd4j.zeros(new int[]{words.size(), truncLen}, 'f');

int[] temp = new int[2];
for(int i=0; i < words.size(); i++){
	word = words.get(i);
	temp[0] = i;

	for(int j=0; j < word.length() && j < truncLen; j++){
		char token = word.charAt(j);

		INDArray vector = Nd4j.zeros(CHARSET.size());
		vector.putScalar(new int[]{CHARSET.indexOf(token)}, 1.0);
		features.put(new INDArrayIndex[]{NDArrayIndex.point(i), 
				NDArrayIndex.all(), NDArrayIndex.point(j)}, vector);

		temp[1] = j;
		featuresMask.putScalar(temp, 1.0);  // Word is present
	}

	int idx = (outcomes[i] ? 0 : 1);
	int lastIdx = Math.min(word.length(), truncLen);
	labels.putScalar(new int[]{i, idx, lastIdx-1}, 1.0);   
	labelsMask.putScalar(new int[]{i, lastIdx-1}, 1.0); 
}

return new DataSet(features, labels, featuresMask, labelsMask);
</pre>

**Comment by AlexDBlack** [comment]
So I'm seeing a few issues here.
First: missing activation functions on the LSTM layer (use "tanh" usually for LSTM layers) and the output layer (use "softmax" for classification)
Second: missing loss function on the output layer. For classification, you want MCXENT (or equivalently negative log likelihood).
Not sure on the learning rate. If that is too high (it's data set specific) you can have problems with convergence. Try lower values if you are still seeing issues after making the other changes.

**Comment by westand** [comment]
Thank you for the response. 

I knowingly stripped off the activation and loss functions for purposes of a sanity check.  When they are included and processed over the full data set (several million examples), a friend using TensorFlow is getting accuracy+20% what I get.  He has provided his network set up and I have tried to mimic it exactly in DL4J. I have tried tuning my hyper-parameters according to many guides and duplicating his TensorFlow initialization. 

** This broad performance gap suggests something is fundamentally wrong in my setup. Therefore to sanity check I was using just 16 training examples and a stupidly bare-bones network (stripping activation and less functions). My question here is that: "given it takes a layer of 128 bLSTM nodes nearly 1000 epochs to overfit just 16 examples with loss nearing 0, is this likely indicative of an error in the iterator? or elsewhere?"

**Comment by westand** [comment]
Ahh. I see now. Gutting the network parameters was a massive de-optimization. With activation and loss functions the cost goes asymptotic towards zero in < 50 epochs with a tiny number of training examples. The performance gap still exists between me and my colleague, but at least I know this sanity check is passed.

**Comment by AlexDBlack** [comment]
The difference could be as simple as something like different default options, or maybe a slight difference between the implementation of weight initialization schemes perhaps.
Happy to look into that more, if you've got code for both cases.
